### PR TITLE
Fixes ofgroup ref_count in the group stats.

### PR DESCRIPTION
**Pull request by pierventre**
Previously, only the rules were counted.
However, ofgroups can reference other ofgroups.

Updates tests that were failing

### CODE REVIEW

File: ofproto/ofproto.c [hunk 1]
```diff
@@ -7608,7 +7655,11 @@ delete_group_start(struct ofproto *ofproto, ovs_version_t version,
     group->being_deleted = true;
 
     /* Mark all the referring groups for deletion. */
+
+    /* Why is needed ? */
```
Comment (charlesmcchan) [inline ofproto/ofproto.c:7658]: Good question. It looks like the code is trying to remove all flows that reference to the group being deleted?

File: include/openvswitch/ofp-group.h [hunk 2]
```diff
@@ -58,6 +58,9 @@ struct ofputil_bucket {
     struct ofpact *ofpacts;     /* Series of "struct ofpact"s. */
     size_t ofpacts_len;         /* Length of ofpacts, in bytes. */
 
+    bool has_groups;            /* 'has_groups' is true if 'ofpacts' contains
```
Comment (charlesmcchan) [inline include/openvswitch/ofp-group.h:61]: Do we actually need an extra field for this? Can it be derived?

File: tests/nsh.at [hunk 3]
```diff
@@ -299,9 +299,9 @@ AT_DATA([flows.txt], [dnl
 ])
 
 AT_DATA([groups.txt], [dnl
-    add group_id=100,type=indirect,bucket=actions=encap(nsh(md_type=1)),set_field:0x1234->nsh_spi,set_field:0x11223344->nsh_c1,group:200
-    add group_id=200,type=indirect,bucket=actions=encap(nsh(md_type=1)),set_field:0x5678->nsh_spi,set_field:0x55667788->nsh_c1,group:300
     add group_id=300,type=indirect,bucket=actions=encap(ethernet),set_field:11:22:33:44:55:66->dl_dst,3
+    add group_id=200,type=indirect,bucket=actions=encap(nsh(md_type=1)),set_field:0x5678->nsh_spi,set_field:0x55667788->nsh_c1,group:300
```
Comment (charlesmcchan) [inline tests/nsh.at:303]: Why do you need to change the order here?

File: tests/ofproto-dpif.at [hunk 4]
```diff
@@ -318,8 +318,8 @@ AT_CLEANUP
 AT_SETUP([ofproto-dpif - group chaining])
 OVS_VSWITCHD_START
 add_of_ports br0 1 10 11
-AT_CHECK([ovs-ofctl -O OpenFlow12 add-group br0 'group_id=1234,type=all,bucket=set_field:192.168.3.90->ip_src,group:123,bucket=output:11'])
 AT_CHECK([ovs-ofctl -O OpenFlow12 add-group br0 'group_id=123,type=all,bucket=output:10'])
+AT_CHECK([ovs-ofctl -O OpenFlow12 add-group br0 'group_id=1234,type=all,bucket=set_field:192.168.3.90->ip_src,group:123,bucket=output:11'])
```
Comment (charlesmcchan) [inline tests/ofproto-dpif.at:322]: Why do you need to change the order here?

File: ofproto/ofproto.c (hunk 1 above)
Comment (pierventre) [reply ofproto/ofproto.c:7658]: Yes - also the comments are wrong.

File: include/openvswitch/ofp-group.h (hunk 2 above)
Comment (pierventre) [reply include/openvswitch/ofp-group.h:61]: This is done to save an iteration over the ofpacts. It is initialized at the creation of the bucket.

File: tests/nsh.at (hunk 3 above)
Comment (pierventre) [reply tests/nsh.at:303]: With the check I made - it is not allowed to reference a group that is not present. This was not enforced before. While it was enforced for flows pointing to a group that does not exist

File: tests/ofproto-dpif.at (hunk 4 above)
Comment (pierventre) [reply tests/ofproto-dpif.at:322]: same

File: tests/nsh.at (hunk 3 above)
Comment (charlesmcchan) [reply tests/nsh.at:303]: We should explain this in the commit log.

Comment (pierventre) [reply tests/nsh.at:303]: sounds good

File: ofproto/ofproto.c [hunk 5]
```diff
@@ -7633,6 +7681,14 @@ delete_groups_start(struct ofproto *ofproto, struct ofproto_group_mod *ogm)
     struct ofgroup *group;
 
     if (ogm->gm.group_id == OFPG_ALL) {
+        /* Unless there is a way to do things in order.
```
Comment (pierventre) [inline ofproto/ofproto.c:7684]: This is what I added

**Review by charlesmcchan — APPROVED** [review]
You probably want to highlight that you "enforce the referenced group being created before the referencing group is created", which was different from the previous behavior, in the commit message.

Also, mentioning something like "update reference count when a group is referenced by other groups" instead of "fixes something" may help the reviewer catch the main idea easier.

**Pull request merged**